apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex

```diff
@@ -98,33 +104,53 @@ defmodule ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes do
         %{walked_text: "", pipe_call: nil, range: nil},
         fn current_char, remaining_text, current_line, current_col, acc ->
           if current_line - 1 == line and current_col - 1 == col do
-            {:ok, pipe_call, call_range} =
-              get_pipe_call(line, col, acc.walked_text, current_char, remaining_text)
-
-            {remaining_text,
-             %{
-               acc
-               | walked_text: acc.walked_text <> current_char,
-                 pipe_call: pipe_call,
-                 range: call_range
-             }}
+            case get_pipe_call(line, col, acc.walked_text, current_char, remaining_text) do
+              {:ok, pipe_call, call_range} ->
+                {remaining_text,
+                 %{
+                   acc
+                   | walked_text: acc.walked_text <> current_char,
+                     pipe_call: pipe_call,
+                     range: call_range
+                 }}
+
+              {:error, :no_pipe_at_selection} ->
+                {remaining_text,
+                 %{
+                   acc
+                   | walked_text: acc.walked_text <> current_char
+                 }}
+            end
           else
             {remaining_text, %{acc | walked_text: acc.walked_text <> current_char}}
           end
         end
       )
 
-    case result do
-      %{pipe_call: nil} ->
+    with {:result, %{pipe_call: pipe_call, range: range}}
+         when not is_nil(pipe_call) and not is_nil(range) <- {:result, result},
+         {:ok, unpiped_text} <- AST.from_pipe(pipe_call) do
+      text_edit = %TextEdit{newText: unpiped_text, range: range}
+      {:ok, text_edit}
+    else
+      {:result, %{pipe_call: nil}} ->
         {:error, :pipe_not_found}
 
-      %{pipe_call: pipe_call, range: range} ->
-        unpiped_text = AST.from_pipe(pipe_call)
-
-        {:ok, %{edited_text: unpiped_text, edit_range: range}}
+      {:error, :invalid_code} ->
+        {:error, :invalid_code}
     end
   end
 
+  defp get_function_call(line, col, head, cur, original_tail) when cur in ["\n", "\r", "\r\n"] do
+    {head, new_cur} = String.split_at(head, -1)
+    get_function_call(line, col - 1, head, new_cur, cur <> original_tail)
+  end
+
+  defp get_function_call(line, col, head, ")", original_tail) do
+    {head, cur} = String.split_at(head, -1)
+    get_function_call(line, col - 1, head, cur, ")" <> original_tail)
+  end
```

Comment:
@axelson here's what caused the tests you added to break: do to how `do_get_function_call` works, when we're at a closing paren, it will jump and try to pipe the next function call. The same for when we're already at the end of a line.

I'm thinking that maybe the correct fix is to track back the text until we reach a "(" character instead (updating indices properly).

I haven't tested, but this one probably breaks if the line ends as ") \n"

WDYT? Any other ideas?

Reply:
Actually, tracing back blind yields other problems when the cursor is in the leading whitespace for the line :(

Reply:
I"ve added a test which demonstrates the ")\s\n" case still breaking

Reply:
> do to how do_get_function_call works, when we're at a closing paren, it will jump and try to pipe the next function call. The same for when we're already at the end of a line.

Do we need to keep the "jumping forward" behavior? What is it buying us?

Reply:
It's more of an artifact on how the "function call finding" code works. for instance on a multi-line function call like:
```
    f(
    x,
    y,
    z)
```

Having the cursor at the beggining of the line containing `x` will still find the function correctly.

However, we can just for the cursor to be, for instance at a "(" character or a non-whitespace character, for instance

Reply:
ah, I see.

FYI, I found another breaking case so I'm working on fixing that.